### DISCUSSION TITLE
Win10 IKEv2 X.509 EAP-TLS connection

### DISCUSSION THREAD

**Discussion by atcapollo** [strongswan/strongswan#1108] [General]
Having managed to get an Android 12 strongswan roadwarrior connection working with certificates. I am now trying to get a Windows 10 roadwarrior configuration working with certificates.

I've followed https://docs.strongswan.org/docs/5.9/interop/windowsClients.html although I am using ipsec.conf and not swanctl.conf

I've generated my own self-signed X.509 CA Cert on the server side (left) - which is also used for the working Android 12 profile.
I've generated X.509 User Certs for the Windows 10 client side (right), and installed them to the Current User - with SAN set to the FQDN of the server (left).
I've installed the X.509 CA Cert to the Windows 10 client. I've checked the fingerprint of the Windows installed CA to that of the CA on the server and both match.

I am getting the following errors on the strongswan server:

```
Jun 25 21:40:53 Zebedee charon: 11[IKE] local host is behind NAT, sending keep alives
Jun 25 21:40:53 Zebedee charon: 11[IKE] remote host is behind NAT
Jun 25 21:40:53 Zebedee charon: 11[IKE] sending cert request for "C=UK, O=strongSwan, CN=Root CA"
Jun 25 21:40:53 Zebedee charon: 11[ENC] generating IKE_SA_INIT response 0 [ SA KE No N(NATD_S_IP) N(NATD_D_IP) CERTREQ N(FRAG_SUP) N(CHDLESS_SUP) N(MULT_AUTH) ]
Jun 25 21:40:53 Zebedee charon: 11[NET] sending packet: from 192.168.3.100[500] to 81.XYZ.XYZ.XYZ[500] (481 bytes)
Jun 25 21:40:53 Zebedee charon: 12[NET] received packet: from 81.XYZ.XYZ.XYZ[4500] to 192.168.3.100[4500] (576 bytes)
Jun 25 21:40:53 Zebedee charon: 12[ENC] parsed IKE_AUTH request 1 [ EF(1/3) ]
Jun 25 21:40:53 Zebedee charon: 12[ENC] received fragment #1 of 3, waiting for complete IKE message
Jun 25 21:40:53 Zebedee charon: 10[NET] received packet: from 81.XYZ.XYZ.XYZ[4500] to 192.168.3.100[4500] (576 bytes)
Jun 25 21:40:53 Zebedee charon: 10[ENC] parsed IKE_AUTH request 1 [ EF(2/3) ]
Jun 25 21:40:53 Zebedee charon: 10[ENC] received fragment #2 of 3, waiting for complete IKE message
Jun 25 21:40:53 Zebedee charon: 08[NET] received packet: from 81.XYZ.XYZ.XYZ[4500] to 192.168.3.100[4500] (544 bytes)
Jun 25 21:40:53 Zebedee charon: 08[ENC] parsed IKE_AUTH request 1 [ EF(3/3) ]
Jun 25 21:40:53 Zebedee charon: 08[ENC] received fragment #3 of 3, reassembled fragmented IKE message (1532 bytes)
Jun 25 21:40:53 Zebedee charon: 08[ENC] parsed IKE_AUTH request 1 [ IDi CERTREQ N(MOBIKE_SUP) CPRQ(ADDR DNS NBNS SRV ADDR6 DNS6 SRV6) SA TSi TSr ]
Jun 25 21:40:53 Zebedee charon: 08[IKE] received cert request for "C=UK, O=strongSwan, CN=Root CA"
Jun 25 21:40:53 Zebedee charon: 08[IKE] received 59 cert requests for an unknown ca
Jun 25 21:40:53 Zebedee charon: 08[CFG] looking for peer configs matching 192.168.3.100[%any]...81.XYZ.XYZ.XYZ[192.168.1.204]
Jun 25 21:40:53 Zebedee charon: 08[CFG] no matching peer config found
Jun 25 21:40:53 Zebedee charon: 08[IKE] peer supports MOBIKE
Jun 25 21:40:53 Zebedee charon: 08[ENC] generating IKE_AUTH response 1 [ N(AUTH_FAILED) ]

```

On Windows 10 the built-in client initiated connection fails:

"IKE authentication credentials are unacceptable"

My ipsec.conf is:

```
conn win10ikev2EAPTLS
     keyexchange=ikev2
     rekey=no
     leftrsasigkey=%cert
     rightrsasigkey=%cert
     leftcert=client1Cert.pem                                          ##Comment note this is not the CA Cert
     left=192.168.3.100
     leftfirewall=yes
     leftsubnet=0.0.0.0/0%
     leftid="C=UK, O=strongSwan, CN=Zebedee"        ##Comment this is the ID for the leftcert
     leftsendcert=always
     right=%any
     rightid="C=UK, O=strongSwan, CN=Dell"
     rightsourceip=10.0.0.3
#     rightcert=client3Cert-dell.pem
     rightsendcert=never
     rightca="C=UK, O=strongSwan, CN=Root CA"
     dpddelay=30
     dpdaction=clear
     auto=add
     eap_identity=%identity
     rightauth=eap-tls

```

Any pointers?
Why is charon requesting what looks to be the CA Cert?

When I configured X.509 for Android (which didn't require EAP-TLS) I never saw any messages requesting the Root CA certificate. Therefore a little confused about what's going on here.

**Comment by atcapollo** [comment]
Root CA Cert installed:
![rootCAwindows](https://user-images.githubusercontent.com/6827078/175791860-667f02b0-44b2-40d4-9d37-edf948047f56.PNG)

User Cert installed - signed by CA above (right ID):
![userCertwindows](https://user-images.githubusercontent.com/6827078/175791869-81e216a2-921a-4b6d-818d-ef0a1c84a3a7.PNG)

**Comment by tobiasbrunner** [comment]
> I've generated X.509 User Certs for the Windows 10 client side (right), and installed them to the Current User - with SAN set to the FQDN of the server (left).

That doesn't really make sense. Why would the client certificate need a SAN that matches the server's FQDN?

> ```
> Jun 25 21:40:53 Zebedee charon: 08[CFG] looking for peer configs matching 192.168.3.100[%any]...81.XYZ.XYZ.XYZ[192.168.1.204]
> Jun 25 21:40:53 Zebedee charon: 08[CFG] no matching peer config found
> ```

You configured `rightid="C=UK, O=strongSwan, CN=Dell"`, which obviously doesn't match `192.168.1.204`.

> leftcert=client1Cert.pem

Why would you configure the client certificate (considering the name) as the server's certificate?

>  Why is charon requesting what looks to be the CA Cert?

It doesn't. It sends a certificate request for certificates issued by that CA. The client does the same.

> When I configured X.509 for Android (which didn't require EAP-TLS) 

Neither does Windows when using machine certificates (see the docs).

> I never saw any messages requesting the Root CA certificate. Therefore a little confused about what's going on here.

There are several aspects that affect whether certificate requests are sent (e.g. the availability of the CA cert, the `rightsendcert` option, or settings in the Android app).

**Comment by atcapollo** [comment]
Thanks for the response Tobias. Comments in line.

Is there a way to see which certificate the windows 10 client is sending?
My sense is it's
 sending the wrong cert. I can select the CA to use on the windows client
but not the specific certificate to send. The Windows machine only has one
client cert from the self signed CA which is the C=UK, O=strongSwan,
CN=Dell cert.

On Mon, 27 Jun 2022, 08:34 Tobias Brunner, ***@***.***> wrote:

> I've generated X.509 User Certs for the Windows 10 client side (right),
> and installed them to the Current User - with SAN set to the FQDN of the
> server (left).
>
> That doesn't really make sense. Why would the client certificate need a
> SAN that matches the server's FQDN?
>
Sorry should have been clearer. The SAN on the server cert (client1) is set
to the server FQDN.

> Jun 25 21:40:53 Zebedee charon: 08[CFG] looking for peer configs matching 192.168.3.100[%any]...81.XYZ.XYZ.XYZ[192.168.1.204]
> Jun 25 21:40:53 Zebedee charon: 08[CFG] no matching peer config found
>
> You configured rightid="C=UK, O=strongSwan, CN=Dell", which obviously
> doesn't match 192.168.1.204.
>

Agree this seems to be an issue, the log seem to show the Windows client
sending the CA identifier. The client cert for 192.168.1.204 the Windows
machine indeed is "C=UK, O=strongSwan, CN=Dell" with the signing CA "C=UK,
O=strongSwan, CN=Root CA" per the screenshots I added.

leftcert=client1Cert.pem
>
> Why would you configure the client certificate (considering the name) as
> the server's certificate?
>
Client1 = server, bad naming...

> Why is charon requesting what looks to be the CA Cert?
>
> It doesn't. It sends a certificate request for certificates issued by that
> CA. The client does the same.
>
> When I configured X.509 for Android (which didn't require EAP-TLS)
>
> Neither does Windows when using machine certificates (see the docs).
>

I did see that, I cannot use machine certificates.

I never saw any messages requesting the Root CA certificate. Therefore a

**Comment by tobiasbrunner** [comment]
> Is there a way to see which certificate the windows 10 client is sending?

Read the log, it doesn't get this far as the identity doesn't match. You basically can't configure a remote identity with Windows clients because they use their physical, private IP address as IKE identity when using EAP authentication.
